Export _NavigationDestinationInfo as public API

Rodsevich opened the pull request:
This exports _NavigationDestinationInfo as public API

Fixes #139038 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

Rodsevich committed:
export _NavigationDestinationBuilder as public API

Rodsevich committed:
Revert "export _NavigationDestinationBuilder as public API"

This reverts commit adc219e3e0b33cd1aebe38dd7ed3bdc75e7df068.

Rodsevich committed:
export _NavigationDestinationInfo as public API

Rodsevich committed:
Revert "Revert "export _NavigationDestinationBuilder as public API""

This reverts commit c8e25dbf7ef7268167057167c3a63d9d99a1a811.

goderbauer commented:
Hello! It is not clear to me what you can build with this class being public. Can you give some examples? Also, it would be good to cover this in a test prevent us from ever making this private again.

Can you also take a look at the failing analyze check above?

Rodsevich committed:
tests + example over NavigationDestinationInfo

Rodsevich committed:
Linking the example with DartPad on NavigationDestinationInfo

Rodsevich committed:
Merge branch 'master' of github.com:flutter/flutter

Rodsevich commented:
Hi @goderbauer! Thanks for your review.
Here is a rudimentary intended NavigationBar UI that can be easily accomplished with that widget being public (and waaay hard [if not impossible] without it)
![image](https://github.com/flutter/flutter/assets/1907844/7d0f3541-7531-4ff2-a21f-58e075f296ee)

The code for every component would be as this:  
```dart
class _CustomDestination extends StatelessWidget {
  const _CustomDestination(this.icon, {super.key});

  final IconData icon;

  Widget build(BuildContext context) {
    final NavigationDestinationInfo info =
        NavigationDestinationInfo.of(context);
    return InkWell(
      onTap: info.onTap,
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        verticalDirection:
            info.index.isEven ? VerticalDirection.down : VerticalDirection.up,
        children: [
          Icon(icon),
          FadeTransition(
              opacity: info.selectedAnimation,
              child: Text(
                  'label ${info.index + 1}/${info.totalNumberOfDestinations}'))
        ],
      ),
    );
  }
}
```

I also added a test in **navigation_bar_test.dart** and this very example in **navigation_bar.3.dart** files. As a separate commit I added the example to the documentation of the published widget.

I have no idea on how to fix the failing analyze check though. Any help on that would be appreciated

Rodsevich committed:
add missing space after flow control statement

Rodsevich committed:
Merge branch 'master' into master

Rodsevich committed:
Merge branch 'master' into master

goderbauer commented on examples/api/lib/material/navigation_bar/navigation_bar.3.dart:
```diff
@@ -0,0 +1,69 @@
+
+import 'package:flutter/material.dart';
+
+/// Flutter code sample for [NavigationBar] with custom destinations that use [NavigationDestinationInfo] for customizing their contents.
```
nit: please wrap this to 80 chars

goderbauer commented on examples/api/lib/material/navigation_bar/navigation_bar.3.dart:
```diff
@@ -0,0 +1,69 @@
+        children: [
+          Icon(icon),
+          FadeTransition(
+              opacity: info.selectedAnimation,
+              child: Text(
+                  'label ${info.index + 1}/${info.totalNumberOfDestinations}'))
```
formatting:

```suggestion
            opacity: info.selectedAnimation,
            child: Text(
               'label ${info.index + 1}/${info.totalNumberOfDestinations}',
             ),
           ),
```

goderbauer reviewed (commented):
Looks like some checks are failing, can you please take a look and fix those? Thanks!

Rodsevich committed:
Update examples/api/lib/material/navigation_bar/navigation_bar.3.dart

Co-authored-by: Michael Goderbauer <goderbauer@google.com>

Rodsevich committed:
Merge branch 'flutter:master' into master

Rodsevich committed:
Merge branch 'master' into master

goderbauer commented:
@Rodsevich Looks like this is still failing some checks. Do you still have plans to come back to this and clean it up?

Rodsevich commented:
I tried fixing it but had no clue on how to debug the issue. Then life
happened and had no more time on finishing it. Is there any chance you can
continue by yourserf? If not, a clue on how to tacxle the issue will be
welcomed.

El vie, 15 de mar de 2024, 14:45, Michael Goderbauer <
***@***.***> escribió:

> @Rodsevich <https://github.com/Rodsevich> Looks like this is still
> failing some checks. Do you still have plans to come back to this and clean
> it up?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/flutter/flutter/pull/139429#issuecomment-2000149408>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAORZBDDA3GMTQTAAYYW7Z3YYMXT7AVCNFSM6AAAAABAEH62D6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAMBQGE2DSNBQHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
